File: app/src/main/java/openeyes/drawalive/seven/openeyes/MainEyes.java
package openeyes.drawalive.seven.openeyes;

import android.Manifest;
import android.content.pm.PackageManager;
import android.os.Build;
import androidx.core.app.ActivityCompat;
import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.AdapterView;
import android.widget.Button;
import android.widget.LinearLayout;


public class MainEyes extends AppCompatActivity {

   // Used to load the 'native-lib' library on application startup.
   static {
      System.loadLibrary("native-lib");
   }

   @Override
   protected void onCreate(Bundle savedInstanceState) {
      super.onCreate(savedInstanceState);
      selectedCameraId = "0";

      if(ActivityCompat.checkSelfPermission(
            this, Manifest.permission.CAMERA
      ) != PackageManager.PERMISSION_GRANTED) {
         ActivityCompat.requestPermissions(this, new String[]{
               Manifest.permission.CAMERA, Manifest.permission.WRITE_EXTERNAL_STORAGE
         }, 1);
      } else { }

      LinearLayout.LayoutParams params;
      LinearLayout panel = new LinearLayout(this);
      params = new LinearLayout.LayoutParams(
            LinearLayout.LayoutParams.MATCH_PARENT,
            LinearLayout.LayoutParams.MATCH_PARENT
      );
      panel.setLayoutParams(params);
      panel.setOrientation(LinearLayout.VERTICAL);

      LinearLayout btnpanel = new LinearLayout(this);
      params = new LinearLayout.LayoutParams(
            LinearLayout.LayoutParams.MATCH_PARENT,
            LinearLayout.LayoutParams.WRAP_CONTENT
      );
      btnpanel.setLayoutParams(params);
      btnpanel.setOrientation(LinearLayout.HORIZONTAL);
      Button button;
      button = new Button(this);
      button.setText("Start");
      button.setOnClickListener(new View.OnClickListener() {
         @Override
         public void onClick(View view) {
            preview.safeCameraOpen(selectedCameraId);
         }
      });
      btnpanel.addView(button);
      button = new Button(this);
      button.setText("Stop");
      button.setOnClickListener(new View.OnClickListener() {
         @Override
         public void onClick(View view) {
            preview.stopPreview();
         }
      });
      btnpanel.addView(button);
      panel.addView(btnpanel);

      btnpanel = new LinearLayout(this);
      params = new LinearLayout.LayoutParams(
            LinearLayout.LayoutParams.MATCH_PARENT,
            LinearLayout.LayoutParams.WRAP_CONTENT
      );
      btnpanel.setLayoutParams(params);
      btnpanel.setOrientation(LinearLayout.HORIZONTAL);
      button = new Button(this);
      button.setText("R+-0");
      button.setOnClickListener(new View.OnClickListener() {
         @Override
         public void onClick(View view) {
            preview.rotateCamera(0);
         }
      });
      btnpanel.addView(button);
      button = new Button(this);
      button.setText("R+90");
      button.setOnClickListener(new View.OnClickListener() {
         @Override
         public void onClick(View view) {
            preview.rotateCamera(90);
         }
      });
      btnpanel.addView(button);
      button = new Button(this);
      button.setText("R-90");
      button.setOnClickListener(new View.OnClickListener() {
         @Override
         public void onClick(View view) {
            preview.rotateCamera(-90);
         }
      });
      btnpanel.addView(button);
      button = new Button(this);
      button.setText("R180");
      button.setOnClickListener(new View.OnClickListener() {
         @Override
         public void onClick(View view) {
            preview.rotateCamera(180);
         }
      });
      btnpanel.addView(button);
      panel.addView(btnpanel);

      btnpanel = new LinearLayout(this);
      params = new LinearLayout.LayoutParams(
            LinearLayout.LayoutParams.MATCH_PARENT,
            LinearLayout.LayoutParams.WRAP_CONTENT
      );
      btnpanel.setLayoutParams(params);
      btnpanel.setOrientation(LinearLayout.HORIZONTAL);
      CameraSpinner spinner = new CameraSpinner(this);
      spinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
         @Override
         public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
            selectedCameraId = (String)spinner.getItemAtPosition(position);
         }

         @Override
         public void onNothingSelected(AdapterView<?> parent) {
            selectedCameraId = "0";
         }
      });
      btnpanel.addView(spinner);
      panel.addView(btnpanel);

      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
         preview = new Preview2(this);
      } else {
         preview = new Preview1(this);
      }
      panel.addView(preview.getImageView());
      this.setContentView(panel);

      // ref: https://developer.android.com/training/scheduling/wakelock.html
      // getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
   }

   @Override
   protected void onPause() {
      super.onPause();
      preview.stopPreviewAndFreeCamera();
   }

   @Override
   protected void onResume() {
      super.onResume();
   }

   @Override
   protected void onDestroy() {
      super.onDestroy();
      preview.stopPreviewAndFreeCamera();
   }

   private native String stringFromJNI();

   private Preview preview;
   private String selectedCameraId;
}

File: app/src/main/kotlin/openeyes/drawalive/seven/openeyes/MainEyes.kt
package openeyes.drawalive.seven.openeyes

import android.Manifest
import android.content.pm.PackageManager
import android.os.Build
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.v4.app.ActivityCompat
import android.widget.Button
import android.widget.LinearLayout

class MainEyes : AppCompatActivity() {

    private var preview: Preview? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        if (ActivityCompat.checkSelfPermission(
                        this, Manifest.permission.CAMERA
                ) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.CAMERA, Manifest.permission.WRITE_EXTERNAL_STORAGE), 1)
        } else {
        }

        var button: Button
        var params: LinearLayout.LayoutParams
        val panel = LinearLayout(this)
        params = LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.MATCH_PARENT
        )
        panel.layoutParams = params
        panel.orientation = LinearLayout.VERTICAL

        val btnpanel = LinearLayout(this)
        params = LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.WRAP_CONTENT
        )
        btnpanel.layoutParams = params
        btnpanel.orientation = LinearLayout.HORIZONTAL
        button = Button(this)
        button.text = "Start"
        button.setOnClickListener { preview!!.safeCameraOpen() }
        btnpanel.addView(button)
        button = Button(this)
        button.text = "Stop"
        button.setOnClickListener { preview!!.stopPreview() }
        btnpanel.addView(button)
        panel.addView(btnpanel)

        val camerabtn = LinearLayout(this)
        params = LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.WRAP_CONTENT
        )
        camerabtn.layoutParams = params
        camerabtn.orientation = LinearLayout.HORIZONTAL
        button = Button(this)
        button.text = "R+-0"
        button.setOnClickListener { preview!!.rotateCamera(0) }
        camerabtn.addView(button)
        button = Button(this)
        button.text = "R+90"
        button.setOnClickListener { preview!!.rotateCamera(90) }
        camerabtn.addView(button)
        button = Button(this)
        button.text = "R-90"
        button.setOnClickListener { preview!!.rotateCamera(-90) }
        camerabtn.addView(button)
        button = Button(this)
        button.text = "R180"
        button.setOnClickListener { preview!!.rotateCamera(180) }
        camerabtn.addView(button)
        panel.addView(camerabtn)

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            preview = Preview2(this)
        } else {
            preview = Preview1(this)
        }
        panel.addView(preview!!.imageView)
        this.setContentView(panel)

        // ref: https://developer.android.com/training/scheduling/wakelock.html
        // getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
    }

    override fun onPause() {
        super.onPause()
        preview!!.stopPreviewAndFreeCamera()
    }

    override fun onResume() {
        super.onResume()
    }

    override fun onDestroy() {
        super.onDestroy()
        preview!!.stopPreviewAndFreeCamera()
    }

    /**
     * A native method that is implemented by the 'native-lib' native library,
     * which is packaged with this application.
     */
    external fun stringFromJNI(): String

    companion object {

        // Used to load the 'native-lib' library on application startup.
        init {
            System.loadLibrary("native-lib")
        }
    }
}
